### PR TITLE
Cache Serailize API

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Adding artifact reservation API. This construct will be used to restrict concurrent execution of multiple instances of a cached task with the same inputs, mitigating the overhead of identical workloads.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Implementing GetOrReserveArtifact and RelaseReservation API calls to manage artifact reservations. The GetOrReserveArtifact call returns the cached artifact if it exists, if not it performs a conditional upsert to claim the artifact reservation. If a task currently holds an artifact reservation, this call is meant to be repeated to update the reservation expiration until completion. Once completed, the task creates artifact using CreateArtifact and releases the reservation with ReleaseReservation.

Waiting on https://github.com/flyteorg/flyteidl/pull/215 to remove ExtendReservation API call and include expiration time and recommended heartbeat interval in GetOrReserveArtifact response.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/267
https://github.com/flyteorg/flyte/issues/420
https://github.com/flyteorg/flyte/issues/872

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/runtime/configs/data_catalog_config.go [hunk 1]
```diff
@@ -4,7 +4,9 @@ package configs
...
-	MetricsScope  string `json:"metrics-scope" pflag:",Scope that the metrics will record under."`
-	ProfilerPort  int    `json:"profiler-port" pflag:",Port that the profiling service is listening on."`
+	StoragePrefix                  string `json:"storage-prefix" pflag:",StoragePrefix specifies the prefix where DataCatalog stores offloaded ArtifactData in CloudStorage. If not specified, the data will be stored in the base container directly."`
+	MetricsScope                   string `json:"metrics-scope" pflag:",Scope that the metrics will record under."`
+	ProfilerPort                   int    `json:"profiler-port" pflag:",Port that the profiling service is listening on."`
+	HeartbeatGracePeriodMultiplier int    `json:"heartbeat-grace-period-multiplier" pflag:",Number of heartbeats before a reservation expires without an extension."`
+	MaxReservationHeartbeatSec     int    `json:"max-reservation-heartbeat-sec" pflag:",The maximum available reservation extension heartbeat interval."`
```
Comment (EngHabu) [inline pkg/runtime/configs/data_catalog_config.go:11]: I would use 
```suggestion
	MaxReservationHeartbeat     config.Duration    `json:"max-reservation-heartbeat" pflag:",The maximum available reservation extension heartbeat interval."`
```

File: boilerplate/flyte/golang_dockerfile/Dockerfile.GoTemplate [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # 
 # TO OPT OUT OF UPDATES, SEE https://github.com/flyteorg/boilerplate/blob/master/Readme.rst
 
-FROM golang:1.17.1-alpine3.14 as builder
+FROM golang:1.13.3-alpine3.10 as builder
```
[the remark this answers is not in the document]
Comment (hamersaw) [reply boilerplate/flyte/golang_dockerfile/Dockerfile.GoTemplate]: I must have mucked this up somewhere. The actual dockerfile in the root was correct, but I have updated this to reflect the current master.

File: pkg/manager/impl/reservation_manager.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+type reservationManager struct {
+	repo                           repositories.RepositoryInterface
+	heartbeatGracePeriodMultiplier time.Duration
+	maxHeartbeatInterval           time.Duration
+	now                            NowFunc
+	systemMetrics                  reservationMetrics
+}
+
+func NewReservationManager(
```
[the remark this answers is not in the document]
Comment (hamersaw) [reply pkg/manager/impl/reservation_manager.go:42]: done

File: pkg/runtime/configs/data_catalog_config.go (hunk 1 above)
Comment (hamersaw) [reply pkg/runtime/configs/data_catalog_config.go]: done

File: pkg/manager/impl/reservation_manager.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	return reservation, nil
+}
+
+func (r *reservationManager) ReleaseReservation(ctx context.Context, request *datacatalog.ReleaseReservationRequest) (*datacatalog.ReleaseReservationResponse, error) {
+	repo := r.repo.ReservationRepo()
+	reservationKey := transformers.FromReservationID(request.ReservationId)
+
+	err := repo.Delete(ctx, reservationKey)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to release reservation: %+v, err: %v", reservationKey, err)
```
[the remark this answers is not in the document]
Comment (hamersaw) [reply pkg/manager/impl/reservation_manager.go]: done

**Review by EngHabu — APPROVED** [review]
👍🏽

File: go.mod [hunk 5]
```diff
@@ -18,3 +18,5 @@ require (
 	gorm.io/driver/postgres v1.1.0
 	gorm.io/gorm v1.21.9
 )
+
+replace github.com/flyteorg/flyteidl => github.com/hamersaw/flyteidl v0.19.26-0.20211103115633-100abab11c51
```
Comment (EngHabu) [inline go.mod:22]: Reminder to revert this when idl is merged

Comment (hamersaw) [reply go.mod]: done

**Review by EngHabu — APPROVED** [review]
